### PR TITLE
reinject BGP routes when local interface address is updated

### PR DESCRIPTION
**Pull request by ishidawataru**
Signed-off-by: ISHIDA Wataru <ishida.wataru@lab.ntt.co.jp>

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -100,6 +100,41 @@ func getEtcdConfig(cfg *calicoapi.CalicoAPIConfig) (etcd.Config, error) {
...
+	}
+	r := routes[0]
+	if r.Gw != nil {
+		return r.Gw, nil
+	}
+	// bgpNexthop can be reached by a connected route
+	return bgpNexthop, nil
+}
+
+func cleanUp() error {
```
Comment (gunjan5) [inline main.go]: can this be called `cleanUpRoutes`? so it's more clear what it's cleaning up

File: main.go [hunk 2]
```diff
@@ -194,6 +231,10 @@ func (s *Server) Serve() {
 	s.t.Go(func() error { return fmt.Errorf("watchKernelRoute: %s", s.watchKernelRoute()) })
 
 	<-s.t.Dying()
+
+	if err := cleanUp(); err != nil {
+		log.Fatal("%s, also failed to clean up routes which we injected: %s", s.t.Err(), err)
```
Comment (gunjan5) [inline main.go]: this should be `Fatalf` else it will print `%s` as it is

File: main.go [hunk 3]
```diff
@@ -789,6 +868,15 @@ func (s *Server) injectRoute(path *bgptable.Path) error {
 		log.Printf("removed route %s from kernel", nlri)
 		return netlink.RouteDel(route)
 	}
+	if !ipip {
+		// Recursive nexthop lookup for GCE environment
```
Comment (gunjan5) [inline main.go]: curious what happens if this was in AWS or some other cloud environment if you happen to know. Would this behavior still be valid?

**Review by gunjan5 — COMMENTED** [review]
few comments, you can squash after addressing those

File: main.go (hunk 3 above)
Comment (ishidawataru) [reply main.go]: Yes, it is also valid for other cloud environments.

The comment was misleading. I updated the comment. PTAL

**Comment by ishidawataru** [comment]
@gunjan5 reflected your comments and squashed. PTAL

**Review by gunjan5 — APPROVED** [review]
LGTM! Thanks @ishidawataru as usual :)

**Comment by ishidawataru** [comment]
@gunjan5 Thanks for merging. When will be the next release of calico-bgp-daemon?

[PR for adding ST for IPIP](https://github.com/projectcalico/calicoctl/pull/1601) is pending, so it'd be great if you release the next version including this commit.

**Comment by gunjan5** [comment]
@ishidawataru I'm planning to do a `v0.2.1` release tomorrow and then we can pull that into calicoctl and then rebase the ST PR

**Pull request merged**